### DISCUSSION TITLE
NuGet packages

### DISCUSSION THREAD

**Discussion by favoyang** [openupm/openupm#553] [Ideas]
Original discussion: https://forum.unity.com/threads/introducing-openupm-upm-registry-with-automatic-build-pipelines.798369/#post-5914997

Nesting DLLs directly into the UPM package is a hacky practice to make a single package just work. But get troubled when two packages included the same DLL. To resolve the conflict, you have to eject (embed) one package manually to the Packages, then remove the DLL. The ideal solution is to have a group of NuGet packages converts to UPM format to allow other packages to reference.

xoofx's UnityNuGet is such a project with limited resources and an automatic approach. These packages are using org.nuget as scope.

https://github.com/worldreaver is also something similar and published to OpenUPM. e.g. https://openupm.com/packages/com.worldreaver.systemmemory/. These packages are using com.worldreaver as scope.

There're other cases like https://openupm.com/packages/jillejr.newtonsoft.json-for-unity/ does for Newtonsoft.Json, by adding more features like custom builds targeting different platforms and link.xml for code stripping.

Notice that though these packages are using their own package namespace, the DLL filenames are the same. So you can only install one of them.

What I may come with is a solution to make it easier to publish NuGet package to OpenUPM.
- It can be something like xoofx does, using an automatic approach. I need more research to see if it can be achieved for adding customization automatically like the newtonsoft.json example above.
- It can be something like worldreaver does use one GitHub repository for one NuGet package, but with a more "official" namespace. This will be less work for OpenUPM to verify those packages and more flexible to add customizations. And we may need to develop a GitHub Action to help track NuGet changes and create PR automatically.

I'm not sure which way to go, probably need some trial-and-error process. When these packages are ready, we can start to promote it to the community, but promoting a good practice may take a long time.

**Comment by bdovaz** [comment]
Any updates @favoyang ??

**Comment by favoyang** [comment]
@bdovaz Nope... haven't got enough time to pick is up.

**Comment by bdovaz** [comment]
Ping! How is this going @favoyang?

I hope you take @xoofx 's approach.

**Comment by favoyang** [comment]
I apologize for the long delay. Our resources are limited. We're currently focusing on the adoption of the [Unity TOS updates](https://forum.unity.com/threads/updates-to-our-terms-of-service-and-new-package-guidelines.999940/). This may happen next year.

**Comment by runette** [comment]
It would be interesting if you could do this is a way that also links to other sources of compiled shared objects - I am particularly thinking of Conda in our case.

I wrote an article about [Unity Conda integration](https://medium.com/runic-software/using-conda-as-a-unity-package-ab2da0386c43) but the current solution is a little hacky.

**Comment by bdominguez** [comment]
Really interesting feature, any news on this?

I really like this approach:

https://github.com/xoofx/UnityNuGet

**Comment by hybridherbst** [comment]
Advantages of xoofx' manual, curated approach that should ideally roll into whatever solution OpenUPM adopts:

- a package is only added if all dependencies are also added
- a package must have a .NET Standard 2 version that is non-preview (that one is shipped to Unity)
- a package must work in Unity
- all packages follow the same naming scheme, `org.nuget.<Name.Of.Package>`
- package updates are released automatically
- does directly fetch from NuGet, no repository required (which would be an issue for all the dotnet packages which come from the giant dotnet repository directly).

I opened an issue (#1976) for an interim solution to this: using xoofx' registry as an uplink registry. This already brings over many of the dotnet dependency packages that are required for a lot of NuGet libraries that target .NET Standard 2.0 (which is required for platform-wide Unity support). Additional packages can be PRed to his registry if needed. And since the naming and structure already feels generally right, future "real" OpenUPM NuGet support could most likely just cut the uplink, replace with a custom solution and all packages would still work exactly the same.

Even if xoofx' registry would suddenly go away, I believe the uplink feature in Verdaccio is implemented in a way that versions are cached, so nothing would break immediately (already fetched versions would still be available).

**Comment by bdovaz** [comment]
In the last week I have contributed several PRs to the xoofx repository:
- Update the project to .NET 5 and refactor / improve certain parts
- Document how to deploy a Docker container
- Add several widely used packages like Microsoft.Extensions. *

For me it is the best solution for openupm

**Comment by favoyang** [comment]
Anyone knows more about the xoofx repository, please help me with https://github.com/xoofx/UnityNuGet/issues/25

**Reply by bdovaz** [reply]
It seems that it's already fixed

**Comment by bdovaz** [comment]
@favoyang How is this going?

I have contributed as I said in several PR and I have to get the last one approved by @xoofx to allow the addition of packages with 4 components.

After all those PRs it is a very solid solution as Nuget integration in UPM.

**Reply by favoyang** [reply]
Working on it. ETA later this month or early next month.

**Reply by favoyang** [reply]
Update: https://github.com/openupm/openupm/issues/1976#issuecomment-831413099
